pkg/runtime/interfaces.go

```diff
@@ -23,64 +23,74 @@ import (
 	"k8s.io/kubernetes/pkg/api/unversioned"
 )
 
-// Codec defines methods for serializing and deserializing API objects.
-type Codec interface {
-	Decoder
-	Encoder
-}
+const (
+	APIVersionInternal    = "__internal"
+	APIVersionUnversioned = "__unversioned"
+)
 
-// Decoder defines methods for deserializing API objects into a given type
-type Decoder interface {
-	Decode(data []byte) (Object, error)
-	// TODO: Remove this method?
-	DecodeToVersion(data []byte, groupVersion unversioned.GroupVersion) (Object, error)
-	DecodeInto(data []byte, obj Object) error
-	// TODO: Remove this method?
-	DecodeIntoWithSpecifiedVersionKind(data []byte, obj Object, groupVersionKind unversioned.GroupVersionKind) error
-
-	DecodeParametersInto(parameters url.Values, obj Object) error
+// Typer retrieves information about an object's group, version, and kind.
+type Typer interface {
+	// ObjectKind returns the version and kind of the provided object, or an
+	// error if the object is not recognized (IsNotRegisteredError will return true).
```

Comment:
Yes - implementing runtime.Object means you can be part of a scheme.  It
does not mean you are part of this scheme.

On Thu, Dec 10, 2015 at 8:49 AM, David Eads <notifications@github.com>
wrote:

> In pkg/runtime/interfaces.go
> <https://github.com/kubernetes/kubernetes/pull/17922#discussion_r47227244>
> :
>
> >
> > -// Decoder defines methods for deserializing API objects into a given type
> > -type Decoder interface {
> > -	Decode(data []byte) (Object, error)
> > -	// TODO: Remove this method?
> > -	DecodeToVersion(data []byte, groupVersion unversioned.GroupVersion) (Object, error)
> > -	DecodeInto(data []byte, obj Object) error
> > -	// TODO: Remove this method?
> > -	DecodeIntoWithSpecifiedVersionKind(data []byte, obj Object, groupVersionKind unversioned.GroupVersionKind) error
> > -
> > -	DecodeParametersInto(parameters url.Values, obj Object) error
> > +// Typer retrieves information about an object's group, version, and kind.
> > +type Typer interface {
> > +	// ObjectKind returns the version and kind of the provided object, or an
> > +	// error if the object is not recognized (IsNotRegisteredError will return true).
>
> Given the new methods on runtime.Object, would I ever end up unrecognized?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/17922/files#r47227244>.
>